src/05-instruction-experiments/server.ts

```diff
@@ -0,0 +1,223 @@
+import chalk from "chalk"
+
+/*
+ * Setup express server, OpenAi client, and config.
+ */
+
+dotenv.config()
+const openai = new OpenAI() // Client will look for OPENAI_API_KEY in your environment during instantiation
+const app = express()
+
+app.use(cors())
+app.use(express.text())
+
+/*
+ * Define the tools that the assistant can use:
+ *
+ * 1. get_user_profile: The user's artsy profile as returned from th "me" query. Will look up user encoded in the x-access-token header.
+ */
+
+const tools: OpenAI.Chat.Completions.ChatCompletionTool[] = [
+  {
+    type: "function",
+    function: {
+      name: "get_user_profile",
+      description: `Get information associated with a user profile on artsy.`,
+      parameters: {
+        type: "object",
+        properties: {
+          token: {
+            type: "string",
+            description:
+              "user's access token, which can be used to fetch user information from artsy.",
+          },
+        },
+      },
+    },
+  },
+]
+
+/*
+ * Define the express server routes
+ */
+
+app.post("/", async (req: Request, res: Response) => {
+  try {
+    res.header("Content-Type", "text/plain")
+
+    console.log(chalk.yellow("Received request body: "), req.body)
+
+    /*
+     * Step 1: Get the response from the assistant, passing in the entire message history.
+     *         We expect the entire message history to come back from the client
+     */
+
+    const messages: OpenAI.Chat.ChatCompletionMessageParam[] = JSON.parse(
+      req.body
+    )
+
+    // get a response from the model, providing the system instructions as the first message.
+    const response = await openai.chat.completions.create({
+      // NOTE: client is unaware of the system message since we don't append it to the messages array.
+      messages: [{ role: "system", content: instructions }, ...messages],
+      model: "gpt-4-turbo",
+      tools,
+    })
+
+    console.log(
+      chalk.yellow("First response from model: "),
+      JSON.stringify(response)
+    )
+
+    const responseMessage = response.choices[0].message
+
+    /*
+     * Step 2: check if the model wants to call a function in response to the user prompt
+     */
+
+    if (responseMessage.tool_calls) {
+      // If the assistant requires a tool call, then we need to make the function call and pass the response back to the assistant
+
+      const toolCalls = responseMessage.tool_calls
+
+      console.log(
+        chalk.yellow("Model requesting to make function call(s): "),
+        toolCalls
+      )
+
+      // extend conversation with assistant's reply. When providing a function call, we need to include the assistant's response in the messages array directly before the response (i.e. order matters).
+      messages.push(responseMessage)
+
+      // The model can request multiple tool calls, so we need to iterate over each one and make the function call.
+      for (const toolCall of toolCalls) {
```

Comment:
**question:** In practice, do we observe the LLM making multiple parallel calls?

Reply:
I haven't tested this, yet, but wanted to leave the option open, as it seemed reasonable that the model could want to use multiple queries/tools to resolve a prompt with just a bit more complexity. But more testing will be needed to determine if this is a YAGTNI situation. 

We may also not want to test beyond the `me` profile, in which  case we can definitely simplify.